mapillm/exact_matching.py

```diff
@@ -0,0 +1,131 @@
+import os
+from mp_api.client import MPRester
+import sys
+
+
+class SuppressOutput:
+    #this is just because I'm annoyed with the print statements and progress bars from mp-api
+    def __enter__(self):
+        self._original_stdout = sys.stdout
+        sys.stdout = open(os.devnull, 'w')
+        self._original_stderr = sys.stderr
+        sys.stderr = open(os.devnull, 'w')
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        sys.stdout.close()
+        sys.stdout = self._original_stdout
+        sys.stderr.close()
+        sys.stderr = self._original_stderr
+
+class ExactMatchingTool():
+  #move to other class when ready
+    def __init__(self):
+        self.mpr = MPRester(os.getenv("MAPI_API_KEY"))
+
+    def _get_specific_query(self, docs, space_group_num=None, density=None, lattice=None):
+        if space_group_num:
+            docs = [doc for doc in docs if doc.symmetry.number == space_group_num]
+        if density:
+            #instead of exact matching density, get density match to the first decimal
+            docs = [doc for doc in docs if round(doc.density, 1) == round(density, 1)]
+        if lattice:
+            docs = [doc for doc in docs if doc.symmetry.crystal_system.value == lattice]
+        return docs
+  
+    def query_material_property(self, formula, desired_prop, space_group_num=None, density=None, lattice=None, one_only=True):
```

Comment:
I am still a little confused about how it'll be implemented in the agent. Is the agent supposed to decide if they will query `one_only` or not? 
If the user/agent should have access to it, the warning of returning `one_only` should be in this function. If this will be a private function and the control will be done in the `get_prop` function, then we don't need the `one_only` argument here